### PR TITLE
Implement global and per-View caching of entries

### PR DESCRIPTION
**Pull request by mrcasual**
This adds a global (GravityKit > Settings > GravityView) and View setting to control caching. This also implements long-lived caching (https://docs.gravitykit.com/article/58-about-gravityview-caching) that we haven't had since... ages.

Ref: #1818.

### CODE REVIEW

File: future/includes/class-gv-settings-view.php [hunk 1]
```diff
@@ -100,6 +101,34 @@ public static function defaults( $detailed = false, $group = null ) {
...
+						esc_html_x( 'Caching is enabled by default in the global [url]GravityView settings[/url]. Adjusting the settings here will allow you to manage caching on a per-View basis.', 'Placeholders inside [] are not to be translated.', 'gk-gravityview' ),
+						[
+							'[url]'  => '<a href="' . SettingsFramework::get_instance()->get_plugin_settings_url( Plugin_Settings::SETTINGS_PLUGIN_ID ) . '&s=1">',
+							'[/url]' => '</a>',
+						]
+					),
+					'show_in_shortcode' => false,
+					'article'           => array(
+						'id'  => '54c67bb6e4b051242988550a',
+						'url' => 'https://docs.gravitykit.com/article/58-about-gravityview-cachinghttps://docs.gravitykit.com/article/58-about-gravityview-caching',
```
Comment (zackkatz) [inline future/includes/class-gv-settings-view.php:119]: Incorrect URL

File: includes/class-cache.php [hunk 2]
```diff
@@ -506,39 +506,45 @@ public function get( $key = null ) {
...
+			array(
+				'key'        => $this->key,
+				'expiration' => $expiration,
+			)
+		);
 
-			if ( ! $transient_was_set && $this->use_cache() ) {
-				gravityview()->log->error( 'Transient was not set for this key: ' . $this->key );
-			}
+		$transient_was_set = set_transient( $this->key, $content, $expiration );
```
Comment (zackkatz) [inline includes/class-cache.php:541]: Should we use our `Utils::set_transient()` instead?

File: future/includes/class-gv-settings-plugin.php [hunk 3]
```diff
@@ -228,6 +230,91 @@ public function add_settings( $plugins_data ) {
...
+
+		if ( has_filter( 'gravityview_cache_time_entries' ) ) {
+			$cache_filters_in_use[] = 'gravityview_cache_time_entries';
+		}
+
+		$cache_settings = [];
+
+		if ( ! empty( $cache_filters_in_use ) ) {
+			$notice = 1 === count( $cache_filters_in_use )
+				? esc_html__( 'Caching settings are being manually overridden by the [filter] filter.', 'gk-gravityview' )
```
Comment (zackkatz) [inline future/includes/class-gv-settings-plugin.php:247]: Let's remove "manually", since we don't know how the code got there…

Comment (zackkatz) [reply future/includes/class-gv-settings-plugin.php:247]: We need `// translators: Don't translate…` message

File: future/includes/class-gv-settings-plugin.php [hunk 4]
```diff
@@ -228,6 +230,91 @@ public function add_settings( $plugins_data ) {
...
+
+		$cache_settings = array_merge( $cache_settings, [
+				array(
+					'id'          => 'caching',
+					'type'        => 'checkbox',
+					'title'       => esc_html__( 'Enable Caching', 'gk-gravityview' ),
+					'description' => strtr(
+						esc_html_x( '[url]Enabling caching[/url] improves performance by reducing the number of queries during page loads. When enabled, you can also specify cache duration for entries.', 'Placeholders inside [] are not to be translated.', 'gk-gravityview' ),
+						[
+							'[url]'  => '<a class="underline" href="https://docs.gravitykit.com/article/58-about-gravityview-caching" rel="noopener noreferrer">',
```
Comment (zackkatz) [inline future/includes/class-gv-settings-plugin.php:291]: Let's do `target="_blank"` with new window hidden text

File: future/includes/class-gv-settings-view.php [hunk 5]
```diff
@@ -100,6 +101,34 @@ public static function defaults( $detailed = false, $group = null ) {
...
 				),
+				'caching'                     => array(
+					'label'             => __( 'Enable Caching', 'gk-gravityview' ),
+					'type'              => 'checkbox',
+					'group'             => 'default',
+					'value'             => gravityview()->plugin->settings->get( 'caching' ),
+					'desc'              => strtr(
+						esc_html_x( 'Caching is enabled by default in the global [url]GravityView settings[/url]. Adjusting the settings here will allow you to manage caching on a per-View basis.', 'Placeholders inside [] are not to be translated.', 'gk-gravityview' ),
+						[
+							'[url]'  => '<a href="' . SettingsFramework::get_instance()->get_plugin_settings_url( Plugin_Settings::SETTINGS_PLUGIN_ID ) . '&s=1">',
```
Comment (zackkatz) [inline future/includes/class-gv-settings-view.php:112]: let's escape the URL just 'cause it's making me nervous

**Review by zackkatz — CHANGES_REQUESTED** [review]
I haven't tested this yet, but the code looks really good. Just a few changes so far.

File: includes/class-cache.php (hunk 2 above)
Comment (mrcasual) [reply includes/class-cache.php:541]: We should, but it's a bigger change since the `Cache` class also does automatic clean-ups/etc.

**Pull request merged**